src/tools/generate-copyright/src/main.rs

```diff
@@ -1,54 +1,114 @@
 use std::io::Write;
-use std::path::PathBuf;
+use std::path::{Path, PathBuf};
 
 use anyhow::Error;
 
+mod cargo_metadata;
+
+/// The entry point to the binary.
+///
+/// You should probably let `bootstrap` execute this program instead of running it directly.
+///
+/// Run `x.py run generate-metadata`
 fn main() -> Result<(), Error> {
-    let dest = env_path("DEST")?;
+    let dest_file = env_path("DEST")?;
+    let out_dir = env_path("OUT_DIR")?;
+    let cargo = env_path("CARGO")?;
     let license_metadata = env_path("LICENSE_METADATA")?;
 
-    let metadata: Metadata = serde_json::from_slice(&std::fs::read(&license_metadata)?)?;
+    let collected_tree_metadata: Metadata =
+        serde_json::from_slice(&std::fs::read(&license_metadata)?)?;
+
+    let root_path = std::path::absolute(".")?;
+    let workspace_paths = [
+        Path::new("./Cargo.toml"),
```

Comment:
I think that we have more workspaces. E.g. recently `src/tools/rustbook` was added as a separate workspace. We should probably somehow pass their list from bootstrap, to avoid forgetting about them? :thinking:

Reply:
I would love to see a central slice in bootstrap listing all workspaces :D

Reply:
Hmm, I took a look and "all workspaces" is not as easy as it sounds :laughing: We currenty have the main workspace, bootstrap workspace, `rustbook` workspace, but then also a bunch of workspaces for individual tools, such as RA. I'm not sure if we can generalize that list for all use-cases. For example, this code probably doesn't want to include the workspace of bootstrap and RA (?).

Reply:
I think this tool should include anything we 'ship' with each 'Rust Release', for some value of 'ship' and some value of 'Rust Release'.

Reply:
https://github.com/rust-lang/rust/pull/128588 dedups the list of workspaces to vendor between the two locations where we run `cargo vendor`. Maybe `generate-copyright` should use the same list of workspaces. That is the set of workspaces that may actually be used by `./x.py dist` and `./x.py test`. Everything else would fail to build on offline systems. For example `library/stdarch` workspace is not meant to be used. Some of it's members are used as dependencies of the standard library, but the rest aren't vendored and bootstrap doesn't expose a way to run the stdarch tests either. Tidy also currently has a list of workspaces for which it checks licenses. That one should probably also be deduped with `cargo vendor` and `generate-copyright`.

Reply:
That sounds reasonable but I'm not sure I know enough about the internal org of this repo to do it. Also I'm on vacation now.

@pietroalbini ?